unittests/test_caching.py

```diff
@@ -17,61 +17,25 @@ class TestCaching:
-
-    @pytest.mark.parametrize(
-        "ahb_expression",
-        [
-            pytest.param("Muss [7] U [8]", id="simple expression"),
-            pytest.param(
-                "Muss [7] U [8] X ([112] O [2343] X [3] U [9]) O ([1] U [2] U ([12] O [13]) O [12 U [222])",
-                id="longer expression",
-            ),
-        ],
-    )
-    def test_ahb_expression_cache_performance(self, clear_caches, ahb_expression: str):
-        # be careful with performance tests. they naturally behave different on different systems
-        calls = 1000
-        time_with_cache = timeit.timeit(
-            lambda: parse_ahb_expression_to_single_requirement_indicator_expressions(
-                ahb_expression, disable_cache=False
-            ),
-            number=calls,
-        )
-
-        time_without_cache = timeit.timeit(
-            lambda: parse_ahb_expression_to_single_requirement_indicator_expressions(
-                ahb_expression, disable_cache=True
-            ),
-            number=calls,
-        )
-
-        avg_time_per_parsing_with_cache = time_with_cache / calls
-        avg_time_per_parsing_without_cache = time_without_cache / calls
-        assert avg_time_per_parsing_with_cache < avg_time_per_parsing_without_cache / 100
-        # a 100-fold performance improvement for 1000 calls of a simple expression seems fair
-        # still, the overall time spent in parsing is not large: the order of magnitude for parsing without cache is 1ms
-
-    def test_ahb_expression_cache_disabled(self, clear_caches, mocker):
-        ahb_expression = "Muss [1] U [2]"
-        parse_spy = mocker.spy(ahb_expr_parser, "parse")
-        for _ in range(100):
-            _ = parse_ahb_expression_to_single_requirement_indicator_expressions(ahb_expression, disable_cache=True)
-        assert parse_spy.call_count == 100
-
-    def test_ahb_expression_cache_sync(self, clear_caches, mocker):
+    def test_ahb_expression_cache_sync(self, mocker):
         ahb_expression = "Muss [3] U [4]"
         parse_spy = mocker.spy(ahb_expr_parser, "parse")
-        for _ in range(100):
-            _ = parse_ahb_expression_to_single_requirement_indicator_expressions(ahb_expression)
+        tree_instances: List[Tree] = []
+        number_of_calls = 100
+        for _ in range(number_of_calls):
+            tree_instance = parse_ahb_expression_to_single_requirement_indicator_expressions(ahb_expression)
+            tree_instances.append(tree_instance)
         parse_spy.assert_called_once_with(ahb_expression)
+        # The following assertion is to make sure that each cached call actually returns a new instance of a lark tree.
+        # We do not want the same instance to be returned and eventually be modified over and over again.
+        number_of_distinct_instances = (
+            len(tree_instances)
+            * len(tree_instances)
+            / len([1 for x, y in product(tree_instances, tree_instances) if x is y])
+        )
```

Comment:
der quotient hier soll also im besten Fall 1 sein oder?
Damit da mit `number_of_calls = 100` steht, `100 * 100/100 = 100` korrekt?

Reply:
ja, wenn du schönere mathe am start hast, freue ich mich. so intuitiv ist das ja nicht :D

Reply:
spontan würde mir nur einfallen `len(tree_instances)**2 / len([1 for x, y in product(tree_instances, tree_instances) if x is y])` aber das verschleiert die Idee nur.
Von daher schon okay.